File: axi-py/packages/agenthub/agenthub/lifecycle.py
"""Agent lifecycle — wake, sleep, and transport management.

Module-level functions that take (hub, session) as arguments.
AgentHub delegates to these — not a god object with methods.

The hub injects SDK factories (make_agent_options, create_client,
disconnect_client) at construction. Lifecycle functions call them
without knowing the concrete SDK types.
"""

from __future__ import annotations

import logging
import os
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any

from opentelemetry import trace

from claudewire.events import ActivityState

if TYPE_CHECKING:
    from agenthub.hub import AgentHub
    from agenthub.types import AgentSession

log = logging.getLogger(__name__)
_tracer = trace.get_tracer(__name__)


# ---------------------------------------------------------------------------
# Pure helpers — no hub dependency
# ---------------------------------------------------------------------------


def is_awake(session: AgentSession) -> bool:
    """True if the session has a connected client."""
    return session.client is not None


def is_processing(session: AgentSession) -> bool:
    """True if the agent's query_lock is currently held."""
    return session.query_lock.locked()


def count_awake(sessions: dict[str, AgentSession]) -> int:
    """Count sessions that have a connected client."""
    return sum(1 for s in sessions.values() if s.client is not None)


def reset_activity(session: AgentSession) -> None:
    """Reset idle tracking and activity state for a new query."""
    session.last_activity = datetime.now(UTC)
    session.idle_reminder_count = 0
    session.activity = ActivityState(phase="starting", query_started=datetime.now(UTC))


# ---------------------------------------------------------------------------
# Sleep
# ---------------------------------------------------------------------------


async def sleep_agent(
    hub: AgentHub, session: AgentSession, *, force: bool = False
) -> None:
    """Disconnect an agent's client and release its scheduler slot.

    If force=False (default), skips if the agent's query_lock is held.
    """
    if not force and session.query_lock.locked():
        log.debug("Skipping sleep for '%s' — query_lock is held", session.name)
        return

    if session.client is None:
        return

    with _tracer.start_as_current_span(
        "sleep_agent",
        attributes={"agent.name": session.name, "agent.force": force},
    ):
        log.info("Sleeping agent '%s'", session.name)
        if session.agent_log:
            session.agent_log.info("SESSION_SLEEP")
        session.bridge_busy = False
        session.transport = None
        await hub.disconnect_client(session.client, session.name)
        session.client = None
        hub.scheduler.release_slot(session.name)
        log.info("Agent '%s' is now sleeping", session.name)


# ---------------------------------------------------------------------------
# Wake
# ---------------------------------------------------------------------------


async def wake_agent(hub: AgentHub, session: AgentSession) -> None:
    """Wake a sleeping agent. Requests a scheduler slot and creates an SDK client.

    On resume failure, retries with a fresh session (no resume).
    Raises ConcurrencyLimitError if no slots are available.
    """
    if is_awake(session):
        return

    # Fail fast if the working directory no longer exists — avoids requesting
    # a scheduler slot and two SDK subprocess attempts that will both fail.
    if session.cwd and not os.path.isdir(session.cwd):
        raise RuntimeError(
            f"Working directory does not exist: {session.cwd}"
        )

    async with hub.wake_lock:
        if is_awake(session):
            return

        await hub.scheduler.request_slot(session.name, timeout=hub.slot_timeout)

        log.info("Waking agent '%s' (session_id=%s)", session.name, session.session_id)

        resume_id = session.session_id
        options = hub.make_agent_options(session, resume_id)

        with _tracer.start_as_current_span(
            "wake_agent",
            attributes={
                "agent.name": session.name,
                "agent.type": session.agent_type or "",
                "agent.resumed": bool(resume_id),
                "agent.cwd": session.cwd or "",
            },
        ):
            try:
                client = await hub.create_client(session, options)
                session.client = client
                log.info(
                    "Agent '%s' is now awake (resumed=%s)", session.name, resume_id
                )
                if session.agent_log:
                    session.agent_log.info(
                        "SESSION_WAKE (resumed=%s)", bool(resume_id)
                    )
                # Successful resume — clear any previous failure marker
                session.last_failed_resume_id = None
            except Exception:
                if resume_id:
                    log.warning(
                        "Failed to resume agent '%s' with session_id=%s, retrying fresh",
                        session.name,
                        resume_id,
                    )
                    options = hub.make_agent_options(session, None)
                    try:
                        client = await hub.create_client(session, options)
                    except Exception:
                        hub.scheduler.release_slot(session.name)
                        raise
                    session.client = client
                    session.session_id = None
                    # Remember the failed session_id so we don't save the same
                    # stale ID when the fresh session returns it in result messages.
                    session.last_failed_resume_id = resume_id
                    log.warning(
                        "Agent '%s' woke with fresh session (previous context lost)",
                        session.name,
                    )
                    if session.agent_log:
                        session.agent_log.info(
                            "SESSION_WAKE (resumed=False, fresh after resume failure)"
                        )
                else:
                    hub.scheduler.release_slot(session.name)
                    raise

        await hub.callbacks.on_wake(session.name)


# ---------------------------------------------------------------------------
# Wake-or-queue
# ---------------------------------------------------------------------------


async def wake_or_queue(
    hub: AgentHub,
    session: AgentSession,
    content: Any,
    metadata: Any = None,
) -> bool:
    """Try to wake an agent. On ConcurrencyLimitError, queue the message.

    Returns True if woken successfully, False if queued or failed.
    """
    from agenthub.types import ConcurrencyLimitError

    try:
        await wake_agent(hub, session)
        return True
    except ConcurrencyLimitError:
        session.message_queue.append((content, metadata))
        position = len(session.message_queue)
        awake = count_awake(hub.sessions)
        log.debug(
            "Concurrency limit hit for '%s', queuing message (position %d, %d awake)",
            session.name,
            position,
            awake,
        )
        return False
    except Exception:
        log.exception("Failed to wake agent '%s'", session.name)
        return False

File: axi-rs/crates/axi/src/lifecycle.rs
//! Agent lifecycle — wake, sleep, and transport management.

use chrono::Utc;
use tracing::{debug, info, warn};

use crate::state::BotState;
use crate::types::{AgentSession, HubError, QueuedMessage};
use crate::activity::ActivityState;

// ---------------------------------------------------------------------------
// Pure helpers
// ---------------------------------------------------------------------------

pub const fn is_awake(session: &AgentSession) -> bool {
    session.awake
}

pub fn is_processing(session: &AgentSession) -> bool {
    session.query_lock.try_lock().is_err()
}

pub fn count_awake(sessions: &std::collections::HashMap<String, AgentSession>) -> usize {
    sessions.values().filter(|s| s.awake).count()
}

pub fn reset_activity(session: &mut AgentSession) {
    session.last_activity = Utc::now();
    session.idle_reminder_count = 0;
    session.activity = ActivityState {
        phase: crate::activity::Phase::Starting,
        query_started: Some(Utc::now()),
        ..Default::default()
    };
}

// ---------------------------------------------------------------------------
// Awaiting input sentinel
// ---------------------------------------------------------------------------

pub async fn post_awaiting_input(state: &BotState, name: &str) {
    if !state.config.show_awaiting_input {
        return;
    }
    let mentions: String = state
        .config
        .allowed_user_ids
        .iter()
        .map(|uid| format!("<@{uid}>"))
        .collect::<Vec<_>>()
        .join(" ");
    crate::frontend::post_system(
        state,
        name,
        &format!("Bot has finished responding and is awaiting input. {mentions}"),
    )
    .await;
}

// ---------------------------------------------------------------------------
// Sleep
// ---------------------------------------------------------------------------

pub async fn sleep_agent(state: &BotState, name: &str, force: bool) {
    let sessions = state.sessions.lock().await;
    let session = match sessions.get(name) {
        Some(s) => s,
        None => return,
    };

    if !force && is_processing(session) {
        debug!("Skipping sleep for '{}' — query_lock is held", name);
        return;
    }

    if !session.awake {
        return;
    }

    info!("Sleeping agent '{}'", name);
    drop(sessions);

    crate::claude_process::disconnect_client(state, name).await;

    let mut sessions = state.sessions.lock().await;
    if let Some(session) = sessions.get_mut(name) {
        session.bridge_busy = false;
        session.awake = false;
    }

    state.scheduler().await.release_slot(name).await;
    info!("Agent '{}' is now sleeping", name);
}

// ---------------------------------------------------------------------------
// Wake
// ---------------------------------------------------------------------------

pub async fn wake_agent(state: &BotState, name: &str) -> Result<(), HubError> {
    {
        let sessions = state.sessions.lock().await;
        let session = sessions
            .get(name)
            .ok_or_else(|| HubError::NotFound(name.to_string()))?;
        if is_awake(session) {
            return Ok(());
        }

        if !session.cwd.is_empty() && !std::path::Path::new(&session.cwd).is_dir() {
            return Err(HubError::Other(format!(
                "Working directory does not exist: {}",
                session.cwd
            )));
        }
    }

    let _wake_lock = state.wake_lock.lock().await;

    // Double-check after acquiring lock
    {
        let sessions = state.sessions.lock().await;
        if let Some(session) = sessions.get(name) {
            if is_awake(session) {
                return Ok(());
            }
        }
    }

    state
        .scheduler()
        .await
        .request_slot(name, state.slot_timeout)
        .await?;

    let resume_id = {
        let sessions = state.sessions.lock().await;
        sessions.get(name).and_then(|s| s.session_id.clone())
    };

    info!("Waking agent '{}' (session_id={:?})", name, resume_id);

    match crate::claude_process::create_client(state, name, resume_id.as_deref()).await {
        Ok(()) => {
            let mut sessions = state.sessions.lock().await;
            if let Some(session) = sessions.get_mut(name) {
                session.awake = true;
                session.last_failed_resume_id = None;
            }
            info!(
                "Agent '{}' is now awake (resumed={:?})",
                name, resume_id
            );
        }
        Err(_) if resume_id.is_some() => {
            warn!(
                "Failed to resume agent '{}' with session_id={:?}, retrying fresh",
                name, resume_id
            );
            if crate::claude_process::create_client(state, name, None).await.is_ok() {
                let mut sessions = state.sessions.lock().await;
                if let Some(session) = sessions.get_mut(name) {
                    session.awake = true;
                    session.last_failed_resume_id = resume_id.clone();
                    session.session_id = None;
                }
                warn!(
                    "Agent '{}' woke with fresh session (previous context lost)",
                    name
                );
            } else {
                state.scheduler().await.release_slot(name).await;
                return Err(HubError::Other(format!(
                    "Failed to create client for agent '{name}'"
                )));
            }
        }
        Err(_) => {
            state.scheduler().await.release_slot(name).await;
            return Err(HubError::Other(format!(
                "Failed to create client for agent '{name}'"
            )));
        }
    }

    crate::frontend::on_wake(state, name).await;
    Ok(())
}

// ---------------------------------------------------------------------------
// Wake-or-queue
// ---------------------------------------------------------------------------

pub async fn wake_or_queue(
    state: &BotState,
    name: &str,
    content: crate::types::MessageContent,
    metadata: Option<serde_json::Value>,
) -> bool {
    match wake_agent(state, name).await {
        Ok(()) => {
            // Agent is awake. Check if it's currently processing (query_lock held).
            // If busy, queue the message — the running task's process_message_queue
            // will pick it up. Spawning a competing task causes race conditions
            // where the second task acquires the lock after the agent has slept.
            let busy = {
                let sessions = state.sessions.lock().await;
                sessions.get(name).is_some_and(is_processing)
            };
            if busy {
                let mut sessions = state.sessions.lock().await;
                if let Some(session) = sessions.get_mut(name) {
                    session
                        .message_queue
                        .push_back(QueuedMessage { content, metadata });
                    let position = session.message_queue.len();
                    debug!(
                        "Agent '{}' is busy, queuing message (position {})",
                        name, position
                    );
                }
                false
            } else {
                true
            }
        }
        Err(HubError::ConcurrencyLimit(_)) => {
            let mut sessions = state.sessions.lock().await;
            if let Some(session) = sessions.get_mut(name) {
                session
                    .message_queue
                    .push_back(QueuedMessage { content, metadata });
                let position = session.message_queue.len();
                debug!(
                    "Concurrency limit hit for '{}', queuing message (position {})",
                    name, position
                );
            }
            false
        }
        Err(e) => {
            warn!("Failed to wake agent '{}': {}", name, e);
            false
        }
    }
}
